### PR TITLE
feat(azure-iot-service): add function to import/export by identity

### PR DESCRIPTION
**Pull request by jebrando**
# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-node/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: service/samples/registry_bulk_identity_sample.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                    //                       console.log('--------------\r\njob ' + jobId + ' status:\r\n--------------\r\n' + result);
+                    //                       var status = result.status;
+                    //                       if (status === "completed") {
+                    //                           clearInterval(interval);
+                    //                       }
+                    //                   }
+                    //               });
+                    //            }, 500);
+                    //        }
+                    //    });
```
Comment (anthonyvercolano) [inline service/samples/registry_bulk_identity_sample.js]: How about no commenting code and have this be an if/else controlled by a command parameter?

File: service/samples/registry_bulk_identity_sample.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    Expiry: expiryDate
+  },
+};
+
+var inputContainerName = 'importcontainer';
+var outputContainerName = 'exportcontainer';
+var deviceFile = 'devices.txt';
+
+blobSvc.createContainerIfNotExists(inputContainerName, function (error) {
+    if(error) {
```
Comment (anthonyvercolano) [inline service/samples/registry_bulk_identity_sample.js]: Let's keep the indenting consistent. 2 spaces.

File: service/samples/registry_bulk_identity_sample.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                          if (error) {
+                              console.error('Could not create import devices: ' + error.message + ' : ' + error.responseBody);
+                          } else {
+                              console.log('--------------\r\nDevices Import Job Identifier:--------------\r\n' + result);
+                              var jobId = result.jobId;
+                              var interval = setInterval(function () {
+                                  /**
+                                   * Uncomment this code to test cancelling a job.
+                                   */
+                                //   registry.cancelJob(jobId, function (error, result) {
```
Comment (anthonyvercolano) [inline service/samples/registry_bulk_identity_sample.js]: No commented code - again, you can use args.  We tried to keep our samples such that they never have to be edited.

Comment (anthonyvercolano) [inline service/samples/registry_bulk_identity_sample.js]: >function (error, result) { [](start = 63, length = 26)

Hope these don't result in stuck jobs.  Do we test canceling these days?

File: service/src/registry.ts [hunk 4]
```diff
@@ -444,6 +491,51 @@ export class Registry {
...
+   * @param {Boolean}     excludeKeys             Boolean indicating whether security keys should be excluded from the exported data.
+   * @param {Function}    [done]                  The optional function to call when the job has been created, with two arguments: an error object if an
+   *                                              an error happened, (null otherwise) and the job status that can be used to track progress of the devices export.
+   * @returns {Promise<Registry.JobStatus> | void} Promise if no callback function was passed, void otherwise.
+   */
+  exportDevicesToBlobByIdentity(outputBlobContainerUri: string, excludeKeys: boolean, done: Callback<Registry.JobStatus>): void;
+  exportDevicesToBlobByIdentity(outputBlobContainerUri: string, excludeKeys: boolean): Promise<Registry.JobStatus>;
+  exportDevicesToBlobByIdentity(outputBlobContainerUri: string, excludeKeys: boolean, done?: Callback<Registry.JobStatus>): Promise<Registry.JobStatus> | void {
+    return callbackToPromise((_callback) => {
+      /* Codes_SRS_NODE_IOTHUB_REGISTRY_16_004: [A ReferenceError shall be thrown if outputBlobContainerUri is falsy] */
```
Comment (anthonyvercolano) [inline service/src/registry.ts]: >Y_16_00 [](start = 38, length = 7)

I don't think actually created new specs for these functions and added them in the devdoc.

File: service/test/_registry_test.js [hunk 5]
```diff
@@ -680,6 +680,114 @@ describe('Registry', function () {
...
+    POST /jobs/create?api-version=<version> HTTP/1.1
+    Authorization: <config.sharedAccessSignature>
+    Content-Type: application/json; charset=utf-8
+    Request-Id: <guid>
+
+    {
+      "type": "import",
+      "inputBlobContainerUri": "<input container Uri given as parameter>",
+      "outputBlobContainerUri": "<output container Uri given as parameter>"
+    }
```
Comment (anthonyvercolano) [inline service/test/_registry_test.js]: This spec doesn't actually say that you have a storage authentication property.  That would be nice.

File: service/test/_registry_test.js [hunk 6]
```diff
@@ -680,6 +680,114 @@ describe('Registry', function () {
...
+    ```
+    POST /jobs/create?api-version=<version> HTTP/1.1
+    Authorization: <config.sharedAccessSignature>
+    Content-Type: application/json; charset=utf-8
+    Request-Id: <guid>
+
+    {
+      "type": "export",
+      "outputBlobContainerUri": "<output container Uri given as parameter>",
+      "excludeKeysInExport": "<excludeKeys Boolean given as parameter>"
```
Comment (anthonyvercolano) [inline service/test/_registry_test.js]: >"excludeKeysInExport": "<excludeKeys Boolean given as parameter> [](start = 6, length = 64)

Here too.  I mean you could just delete this requirement, but, that would be disappointing.

File: service/samples/registry_bulk_identity_sample.js (hunk 1 above)
Comment (zolvarga) [reply service/samples/registry_bulk_identity_sample.js]: Or create another clean sample. (registry_bulk_identity_export and _import)

---
In reply to: [388026474](https://github.com/Azure/azure-iot-sdk-node/pull/709#discussion_r388026474) [](ancestors = 388026474)

File: service/src/registry.ts (hunk 4 above)
Comment (jebrando) [reply service/src/registry.ts]: I will update this

---
In reply to: [388029843](https://github.com/Azure/azure-iot-sdk-node/pull/709#discussion_r388029843) [](ancestors = 388029843)

File: service/samples/registry_bulk_identity_sample.js (hunk 1 above)
Comment (jebrando) [reply service/samples/registry_bulk_identity_sample.js]: I just copied the existing sample for this sample.  Let's have a conversation about this offline.

---
In reply to: [388035333](https://github.com/Azure/azure-iot-sdk-node/pull/709#discussion_r388035333) [](ancestors = 388035333,388026474)

File: service/samples/registry_bulk_identity_sample.js (hunk 2 above)
Comment (jebrando) [reply service/samples/registry_bulk_identity_sample.js]: Same as the below comment, we can talk about changing both the existing sample and this one.  I will close this comment

---
In reply to: [388026930](https://github.com/Azure/azure-iot-sdk-node/pull/709#discussion_r388026930) [](ancestors = 388026930)

File: service/samples/registry_bulk_identity_sample.js (hunk 3 above)
Comment (jebrando) [reply service/samples/registry_bulk_identity_sample.js]: I am unaware of any testing for job cancellation.

---
In reply to: [388027734](https://github.com/Azure/azure-iot-sdk-node/pull/709#discussion_r388027734) [](ancestors = 388027734)

File: service/samples/registry_bulk_export_sample.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                   if (error) {
+                       console.error('Could not create output container: ' + error.message);
+                   } else {
+                       var outputSasToken = blobSvc.generateSharedAccessSignature(outputContainerName, null, outputSharedAccessPolicy);
+                       var outputSasUrl = blobSvc.getUrl(outputContainerName, null, outputSasToken);
+
+                    /**
+                     * There can only be one active job at a time, therefore, you can uncomment the export section and comment the import
+                     * session or vice versa depending on what part of the code you would like to test.
+                     */
```
Comment (anthonyvercolano) [inline service/samples/registry_bulk_export_sample.js]: Nit: Shift to the right two spaces.

File: service/src/registry.ts [hunk 8]
```diff
@@ -401,6 +401,54 @@ export class Registry {
...
+  importDevicesFromBlobByIdentity(inputBlobContainerUri: string, outputBlobContainerUri: string, done: Callback<Registry.JobStatus>): void;
+  importDevicesFromBlobByIdentity(inputBlobContainerUri: string, outputBlobContainerUri: string): Promise<Registry.JobStatus>;
+  importDevicesFromBlobByIdentity(inputBlobContainerUri: string, outputBlobContainerUri: string, done?: Callback<Registry.JobStatus>): Promise<Registry.JobStatus> | void {
+    return callbackToPromise((_callback) => {
+      /* Codes_SRS_NODE_IOTHUB_REGISTRY_07_001: [A ReferenceError shall be thrown if importBlobContainerUri is falsy] */
+      if (!inputBlobContainerUri) throw new ReferenceError('inputBlobContainerUri cannot be falsy');
+      /* Codes_SRS_NODE_IOTHUB_REGISTRY_07_002: [A ReferenceError shall be thrown if exportBlobContainerUri is falsy] */
+      if (!outputBlobContainerUri) throw new ReferenceError('outputBlobContainerUri cannot be falsy');
+
+      /*SRS_NODE_IOTHUB_REGISTRY_07_003: [The `importDevicesFromBlobByIdentity` method shall construct an HTTP request using information supplied by the caller, as follows:
```
Comment (anthonyvercolano) [inline service/src/registry.ts]: >SRS [](start = 8, length = 3)

Nit: One blank space followed by Codes_

File: service/src/registry.ts [hunk 9]
```diff
@@ -444,6 +492,52 @@ export class Registry {
...
+   * @returns {Promise<Registry.JobStatus> | void} Promise if no callback function was passed, void otherwise.
+   */
+  exportDevicesToBlobByIdentity(outputBlobContainerUri: string, excludeKeys: boolean, done: Callback<Registry.JobStatus>): void;
+  exportDevicesToBlobByIdentity(outputBlobContainerUri: string, excludeKeys: boolean): Promise<Registry.JobStatus>;
+  exportDevicesToBlobByIdentity(outputBlobContainerUri: string, excludeKeys: boolean, done?: Callback<Registry.JobStatus>): Promise<Registry.JobStatus> | void {
+    return callbackToPromise((_callback) => {
+      /* Codes_SRS_NODE_IOTHUB_REGISTRY_07_004: [A ReferenceError shall be thrown if outputBlobContainerUri is falsy] */
+      if (!outputBlobContainerUri) throw new ReferenceError('outputBlobContainerUri cannot be falsy');
+
+      /*Codes_SRS_NODE_IOTHUB_REGISTRY_07_005: [** The `exportDeviceToBlob` method shall construct an HTTP request using information supplied by the caller, as follows:
```
Comment (anthonyvercolano) [inline service/src/registry.ts]: >/*Codes_SRS_NODE_IOTHUB_REGISTRY_07_005 [](start = 6, length = 39)

Nit: Blank space before Codes.

**Comment by anthonyvercolano** [comment]
I think for now you're just going to have to disable the e2e tests to get through the gate.

**Comment by anthonyvercolano** [comment]
Also could you squash this?

**Pull request merged**